src/Features/Core/Portable/Workspace/BackgroundCompiler.cs

```diff
@@ -40,141 +54,112 @@ public BackgroundCompiler(Workspace workspace)
 
         public void Dispose()
         {
-            if (_workspace != null)
+            _disposalCancellationSource.Cancel();
+            _cancellationSeries.Dispose();
+
+            lock (_gate)
             {
-                CancelBuild(releasePreviousCompilations: true);
+                _mostRecentCompilations.Clear();
+            }
 
+            if (_workspace != null)
+            {
                 _workspace.DocumentClosed -= OnDocumentClosed;
                 _workspace.DocumentOpened -= OnDocumentOpened;
                 _workspace.WorkspaceChanged -= OnWorkspaceChanged;
 
-                _workspace = null!;
+                _workspace = null;
             }
         }
 
         private void OnDocumentOpened(object? sender, DocumentEventArgs args)
-            => Rebuild(args.Document.Project.Solution, args.Document.Project.Id);
+            => Rebuild();
 
         private void OnDocumentClosed(object? sender, DocumentEventArgs args)
-            => Rebuild(args.Document.Project.Solution, args.Document.Project.Id);
+            => Rebuild();
 
         private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs args)
+            => Rebuild();
+
+        private void Rebuild()
         {
-            switch (args.Kind)
-            {
-                case WorkspaceChangeKind.SolutionCleared:
-                case WorkspaceChangeKind.SolutionAdded:
-                case WorkspaceChangeKind.SolutionRemoved:
-                    CancelBuild(releasePreviousCompilations: true);
-                    break;
-
-                case WorkspaceChangeKind.SolutionChanged:
-                case WorkspaceChangeKind.ProjectRemoved:
-                    if (args.NewSolution.ProjectIds.Count == 0)
-                    {
-                        // Close solution no longer triggers a SolutionRemoved event,
-                        // so we need to make an explicitly check for ProjectRemoved event.
-                        CancelBuild(releasePreviousCompilations: true);
-                    }
-                    else
-                    {
-                        Rebuild(args.NewSolution);
-                    }
-
-                    break;
-
-                default:
-                    Rebuild(args.NewSolution, args.ProjectId);
-                    break;
-            }
+            // Stop any work on the current batch and create a token for the next batch.
+            var nextToken = _cancellationSeries.CreateNext();
+            _workQueue.AddWork(nextToken);
```

Comment:
Much simpler eventing model.  On any event we always just cancel whatever work we're doing and start the next batch.